internal/backends/rust/rust.go

```diff
@@ -0,0 +1,251 @@
+		util.Die("crates.io: %s", err)
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case 200:
+		break
+	case 404:
+		return api.PkgInfo{}
+	default:
+		util.Die("crates.io: HTTP status %d", resp.StatusCode)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	var crateInfo crateInfoResult
+	if err := json.Unmarshal(body, &crateInfo); err != nil {
+		util.Die("crates.io: %s", err)
+	}
+
+	return crateInfo.toPkgInfo()
+}
+
+func listSpecfile() map[api.PkgName]api.PkgSpec {
+	contentsB, err := ioutil.ReadFile("Cargo.toml")
+	if err != nil {
+		util.Die("Cargo.toml: %s", err)
+	}
+
+	var specfile cargoToml
+	err = toml.Unmarshal(contentsB, &specfile)
+	if err != nil {
+		util.Die("Cargo.toml: %s", err)
+	}
+
+	packages := make(map[api.PkgName]api.PkgSpec)
+	for name, dependency := range specfile.Dependencies {
+		var spec api.PkgSpec
+		switch value := dependency.(type) {
+		case string:
+			spec = api.PkgSpec(value)
+
+		case map[string]interface{}:
+			found := false
+			for _, key := range []string{"version", "git", "path"} {
+				specStr, ok := value[key].(string)
+				if !ok {
+					continue
+				}
+
+				spec = api.PkgSpec(specStr)
+				found = true
+				break
+			}
+
+			if !found {
+				util.Die("Cargo.toml: could not determine spec for dependecy %q", name)
+			}
+
+		default:
+			util.Die("Cargo.toml: unexpected dependency format %q", name)
+		}
+
+		packages[api.PkgName(name)] = spec
+
+	}
+
+	return packages
+}
+
+func listLockfile() map[api.PkgName]api.PkgVersion {
+	contentsB, err := ioutil.ReadFile("Cargo.lock")
+	if err != nil {
+		util.Die("Cargo.lock: %s", err)
+	}
+
+	var lockfile cargoLock
+	err = toml.Unmarshal(contentsB, &lockfile)
+	if err != nil {
+		util.Die("Cargo.lock: %s", err)
+	}
+
+	packages := make(map[api.PkgName]api.PkgVersion)
+	for _, pkg := range lockfile.Packages {
+		packages[api.PkgName(pkg.Name)] = api.PkgVersion(pkg.Version)
+	}
+
+	return packages
+}
+
+// RustBackend is a UPM backend for Rust that uses Cargo.
+var RustBackend = api.LanguageBackend{
+	Name:             "rust",
+	Specfile:         "Cargo.toml",
+	Lockfile:         "Cargo.lock",
+	FilenamePatterns: []string{"*.rs"},
+	GetPackageDir: func() string {
+		return "target"
+	},
+	Search: search,
+	Info:   info,
+	Add: func(pkgs map[api.PkgName]api.PkgSpec, projectName string) {
+		if !util.Exists("Cargo.toml") {
+			util.RunCmd([]string{"cargo", "init", "."})
+		}
+		cmd := []string{"cargo", "add"}
+		for name, spec := range pkgs {
+			arg := string(name)
+			if spec != "" {
+				arg += "@" + string(spec)
+			}
+			cmd = append(cmd, arg)
+		}
+		util.RunCmd(cmd)
+	},
+	Remove: func(pkgs map[api.PkgName]bool) {
+		cmd := []string{"cargo", "rm"}
+		for name := range pkgs {
+			cmd = append(cmd, string(name))
+		}
+		util.RunCmd(cmd)
+	},
+	Lock: func() {
+		// Lock file is updated at build time
+	},
+	Install: func() {
+		// Dependencies are installed at build time
+	},
+	ListSpecfile: listSpecfile,
+	ListLockfile: listLockfile,
```

Comment:
Added a test for the parsing part. Hard to test the other part since it wants to read "Cargo.toml" in the current directory haha

Reply:
global state!!!!! (shakes fist)

in the future, the right way to test these kinds of things is to plumb in a https://pkg.go.dev/io/fs#FS and use its `.Open()` instead of `os.Open`. but that requires a massive refactor, so meh.